netbsd/NetBSDProcessList.c

```diff
@@ -0,0 +1,328 @@
+
+   for (int i = 0; i < count; i++) {
+      const struct kinfo_proc2* kproc = &kprocs[i];
+
+      bool preExisting = false;
+      Process* proc = ProcessList_getProcess(&this->super, kproc->p_pid, &preExisting, NetBSDProcess_new);
+
+      proc->show = ! ((hideKernelThreads && Process_isKernelThread(proc)) || (hideUserlandThreads && Process_isUserlandThread(proc)));
+
+      if (!preExisting) {
+         proc->ppid = kproc->p_ppid;
+         proc->tpgid = kproc->p_tpgid;
+         proc->tgid = kproc->p_pid;
+         proc->session = kproc->p_sid;
+         proc->tty_nr = kproc->p_tdev;
+         proc->pgrp = kproc->p__pgid;
+         proc->isKernelThread = proc->pgrp == 0;
+         proc->isUserlandThread = proc->pid != proc->tgid;
+         proc->st_uid = kproc->p_uid;
+         proc->starttime_ctime = kproc->p_ustart_sec;
+         Process_fillStarttimeBuffer(proc);
+         proc->user = UsersTable_getRef(this->super.usersTable, proc->st_uid);
+         ProcessList_add(&this->super, proc);
+
+         NetBSDProcessList_updateProcessName(this->kd, kproc, proc);
+      } else {
+         if (settings->updateProcessNames) {
+            NetBSDProcessList_updateProcessName(this->kd, kproc, proc);
+         }
+      }
+
+      proc->m_virt = kproc->p_vm_vsize;
+      proc->m_resident = kproc->p_vm_rssize;
+      proc->percent_mem = (proc->m_resident * pageSizeKB) / (double)(this->super.totalMem) * 100.0;
+      proc->percent_cpu = CLAMP(getpcpu(kproc), 0.0, this->super.cpuCount * 100.0);
+      proc->nlwp = kproc->p_nlwps;
+      proc->nice = kproc->p_nice - 20;
+      proc->time = 100 * (kproc->p_rtime_sec + ((kproc->p_rtime_usec + 500000) / 1000000));
+      proc->priority = kproc->p_priority - PZERO;
+
+      struct kinfo_lwp* klwps = kvm_getlwps(this->kd, kproc->p_pid, kproc->p_paddr, sizeof(struct kinfo_lwp), &nlwps);
+
+      switch (kproc->p_realstat) {
+      case SIDL:     proc->state = 'I'; break;
+      case SACTIVE:
+         // We only consider the first LWP with a one of the below states.
+         for (int j = 0; j < nlwps; j++) {
+            if (klwps) {
+               switch (klwps[j].l_stat) {
+               case LSONPROC: proc->state = 'P'; break;
+               case LSRUN:    proc->state = 'R'; break;
+               case LSSLEEP:  proc->state = 'S'; break;
+               case LSSTOP:   proc->state = 'T'; break;
+               default:       proc->state = '?';
+               }
+               if (proc->state != '?')
+                  break;
+            } else {
+               proc->state = '?';
+               break;
+            }
+         }
+         break;
+      case SSTOP:    proc->state = 'T'; break;
+      case SZOMB:    proc->state = 'Z'; break;
+      case SDEAD:    proc->state = 'D'; break;
+      default:       proc->state = '?';
+      }
+
+      this->super.totalTasks++;
+      // SRUN ('R') means runnable, not running
+      if (proc->state == 'P') {
+         this->super.runningTasks++;
+      }
```

Comment:
This likely also should count the running kernel threads here.

Reply:
We set the `proc->state` to `'P'` based on the status of the running thread, this is done a few lines above. Does that answer your question about counting the running kernel threads?

Reply:
I'm missing a reference to `this->super.this->super.runningTasks++;` for all processes marked as kernel threads (helper threads of the kernel, not necessarily running).

Reply:
I looked up FreeBSD / OpenBSD code and added a section above this to count the Kernel / User threads. Let me know if that works.

Should be available in 68109ea